### DISCUSSION TITLE
week4 作業問題 關於串接api

### DISCUSSION THREAD

**Discussion by backas36** [Lidemy/forum#111] [程式導師實驗計畫] [answered]
這邊是 week4 hw1 我提交的程式碼，原本沒什麼疑惑，但自己在複習的時候，突然想到幾個問題：

- 為什麼 data 要宣告在 try 區塊的 前面，因為試過讓 ``` let data = JSON.parse(body) ``` 放在 try { } 裡面，還是一樣跑得起來，所以很疑惑先宣告是為了什麼呢？
- 另外迴圈也是，如果放到 try 區塊裡面，程式一樣是可以執行，只是不知道放在 try 區塊外面跟裡面的用意是什麼？

```js
request('https://lidemy-book-store.herokuapp.com/books?_limit=10', (err,res,body)=>{
    if(err){
      console.log(err)
      return
    }

    let data
    try{
      data = JSON.parse(body)
    } catch (error) {
      console.log(error)
    }
    for (let i = 0; i < data.length; i++) {
        console.log(`${data[i].id} ${data[i].name}`)
    }
  }
)
```

先謝謝大家囉！

**Comment by allenliaooooooo** [comment, marked as answer]
`data` 要在 try...catch 外面宣告是因為 let 宣告的是區塊變數，離開了 try...catch 區塊就讀不到 `data` 這個變數了
迴圈不要放在 try..catch 裡面是因為，這樣你會不知道是迴圈錯誤還是 `JSON.parse()` 錯誤

**Reply by backas36** [reply]
了解。
當data宣告在try裡面的話，日後要新增其他功能需要 data 的時候，就會有錯誤！
所以必須在 try 外面宣告，這樣才可以被訪問！
感謝！
